allennlp/nn/beam_search.py

```diff
@@ -25,6 +27,365 @@
+        return torch.gather(_log_probs, 1, selected_indices), selected_indices, state
+
+
+@Sampler.register("top-p")
+class TopPSampler(Sampler):
+    """
+    Represents a `Sampler` which redistributes the probability mass function among
+    the top choices with a cumulative probability of at least `p` then selects from that subset
+    `p` if minimum cumulative probability of highest-probability options that the returned
+    choice will be selected from `temperature` modules the probabilitis of the selected tokens.
+    A `temperature` below 1.0 produces a sharper probability distribution and a `temperature`
+    above 1.0 produces a flatter probability distribution.
+    Registered as a `Sampler` with name "top-p".
+    """
+
+    def __init__(
+        self,
+        p: float = 0.9,
+        temperature: float = 1.0,
+        filter_val: float = -float("inf"),
+        with_replacement: bool = False,
+    ):
+        if not isinstance(p, float) or p < 0.0 or p > 1.0:
+            raise ValueError("p must be a non-negative float no greater than 1.0")
+        self.p = p
+        self.temperature = temperature or 1.0
+        self.filter_val = min_value_of_dtype(torch.float)
+        self.with_replacement = with_replacement
+
+    @overrides
+    def sample_nodes(
+        self, log_probs: torch.Tensor, per_node_beam_size: int, state: StateType
+    ) -> Tuple[torch.Tensor, torch.Tensor, StateType]:
+
+        # First apply temperature coefficient:
+        if self.temperature != 1.0:
+            _log_probs = torch.nn.functional.log_softmax(log_probs / self.temperature, dim=-1)
+        else:
+            _log_probs = log_probs
+
+        # Sort the probabilities to hi ghest-first, then find the cumulative sum accross those
+        log_probs_descending, sorting_indices = torch.sort(_log_probs, descending=True)
+        probabilities_descending = torch.nn.functional.softmax(log_probs_descending, dim=-1)
+        probabilities_summed = torch.cumsum(probabilities_descending, dim=-1)
+
+        # When the cumulative sum reaches p, replace all remaining with `filter_val`
+        filtered_indices = probabilities_summed >= self.p
+
+        # We want to include the firt index where probabilities_summes >= p, so we shift over one
+        filtered_indices[..., 1:] = filtered_indices[..., :-1].clone()
+        filtered_indices[..., 0] = 0
+
+        for row, sort_indices, filter_vals in zip(_log_probs, sorting_indices, filtered_indices):
+            filter_idx = sort_indices[filter_vals]
+            row[filter_idx] = self.filter_val
+
+        filtered_probabilities = torch.nn.functional.softmax(_log_probs, dim=-1)
+        print(filtered_probabilities)
+
+        # Here we sample from the filtered distribution
+        selected_indices = torch.multinomial(
+            filtered_probabilities, per_node_beam_size, replacement=self.with_replacement
+        )
```

Comment:
I think we could avoid a loop by doing something like this. This also avoids modifying `_log_probs` in place (which would modify `log_probs` if temperature was 1.0).

```suggestion
        # Create a mask for filtering out probabilities that don't make the top `p`.
        # shape: (batch_size, num_classes)
        exclusion_mask = probabilities_summed > self.p

        # Now re-normalized the included log probs.
        # shape: (batch_size, num_classes)
        log_probs_descending[exclusion_mask] = min_value_of_dtype(log_probs.dtype)
        # shape: (batch_size, num_classes)
        filtered_probabilities = torch.nn.functional.softmax(log_probs_descending, dim=-1)

        # Sample from the re-normalized subset.
        # NOTE: These indices are not indices into `log_probs`, they are indices into `log_probs_descending`.
        # shape: (batch_size, per_node_beam_size)
        sampled_indices = torch.multinomial(filtered_probabilities, per_node_beam_size, replacement=False)

        # Convert `sampled_indices` back to indices in the original `log_probs` tensor.
        # shape: (batch_size, per_node_beam_size)
        selected_indices = sorting_indices.gather(-1, sampled_indices)
```